File: test/lua/utility/numeric/multi_index.lua
#!/usr/bin/env halmd
--
-- Copyright © 2014 Nicolas Höft
-- Copyright © 2014 Felix Höfling
--
-- This file is part of HALMD.
--
-- HALMD is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
--

local halmd = require("halmd")
halmd.io.log.open_console()

-- grab modules
local log = halmd.io.log
local numeric = halmd.utility.numeric

function test(dims)
    local ndim = 1
    for d = 1, #dims do
        ndim = ndim * dims[d]
    end
    log.info("total number of points: %d", ndim)

    for i = 1, ndim do
        local index = numeric.offset_to_multi_index(i, dims)
        local offset = numeric.multi_index_to_offset(index, dims)
        log.info("offset: %d, index: %d %d %d %d", offset, index[1], index[2], index[3], index[4])
        assert(i == offset)
    end
end

test({2, 1, 3, 3})
test({1})

File: test/unit/utility/multi_index.cpp
/*
 * Copyright © 2012  Peter Colberg
 *
 * This file is part of HALMD.
 *
 * HALMD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 */

#define BOOST_TEST_MODULE index
#include <boost/test/unit_test.hpp>

#include <boost/array.hpp>

#include <halmd/utility/multi_index.hpp>
#include <test/tools/ctest.hpp>

using namespace halmd;

namespace std {

/**
 * Output boost::array to stream for BOOST_CHECK_EQUAL.
 */
template <typename T, size_t N>
ostream& operator<<(ostream& os, boost::array<T, N> const& index)
{
    for (size_t i = 0; i + 1 < N; ++i) {
        os << index[i] << " ";
    }
    return os << index[N - 1];
}

} // namespace std

BOOST_AUTO_TEST_CASE( multi_index_1_dim )
{
    boost::array<size_t, 1> index = {{ 17 }};
    boost::array<size_t, 1> dims = {{ 19 }};
    size_t offset = 17;
    BOOST_CHECK_EQUAL( multi_index_to_offset(index, dims), offset );
    BOOST_CHECK_EQUAL( offset_to_multi_index(offset, dims), index );
}

BOOST_AUTO_TEST_CASE( multi_index_2_dim )
{
    boost::array<size_t, 2> index = {{ 17, 11 }};
    boost::array<size_t, 2> dims = {{ 19, 13 }};
    size_t offset = 17 + 11 * 19;
    BOOST_CHECK_EQUAL( multi_index_to_offset(index, dims), offset );
    BOOST_CHECK_EQUAL( offset_to_multi_index(offset, dims), index );
}

BOOST_AUTO_TEST_CASE( multi_index_3_dim )
{
    boost::array<size_t, 3> index = {{ 11, 17, 13 }};
    boost::array<size_t, 3> dims = {{ 19, 29, 23 }};
    size_t offset = 11 + 17 * 19 + 13 * (19 * 29);
    BOOST_CHECK_EQUAL( multi_index_to_offset(index, dims), offset );
    BOOST_CHECK_EQUAL( offset_to_multi_index(offset, dims), index );
}

BOOST_AUTO_TEST_CASE( multi_index_4_dim )
{
    boost::array<size_t, 4> index = {{ 11, 79, 17, 13 }};
    boost::array<size_t, 4> dims = {{ 19, 131, 29, 23 }};
    size_t offset = 11 + 79 * 19 + 17 * (19 * 131) + 13 * (19 * 131 * 29);
    BOOST_CHECK_EQUAL( multi_index_to_offset(index, dims), offset );
    BOOST_CHECK_EQUAL( offset_to_multi_index(offset, dims), index );
}

BOOST_AUTO_TEST_CASE( multi_index_5_dim )
{
    boost::array<size_t, 5> index = {{ 11, 79, 127, 17, 13 }};
    boost::array<size_t, 5> dims = {{ 19, 131, 137, 29, 23 }};
    size_t offset = 11 + 79 * 19 + 127 * (19 * 131) + 17 * (19 * 131 * 137) + 13 * (19 * 131 * 137 * 29);
    BOOST_CHECK_EQUAL( multi_index_to_offset(index, dims), offset );
    BOOST_CHECK_EQUAL( offset_to_multi_index(offset, dims), index );
}
